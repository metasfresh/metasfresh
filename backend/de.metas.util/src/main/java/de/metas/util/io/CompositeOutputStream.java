package de.metas.util.io;

import com.google.common.collect.ImmutableList;
import de.metas.util.Check;
import lombok.NonNull;

import java.io.File;
import java.io.IOException;
import java.io.OutputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Set;

/**
 * Groups multiple output streams and write to all of them in parallel.
 */
public class CompositeOutputStream extends OutputStream
{
	private final ImmutableList<OutputStream> outputStreams;

	public CompositeOutputStream(@NonNull final ImmutableList<OutputStream> outputStreams)
	{
		this.outputStreams = Check.assumeNotEmpty(outputStreams, "outputStreams cannot be empty");
	}

	public static OutputStream ofList(final List<OutputStream> outputStreams)
	{
		if (outputStreams.isEmpty())
		{
			throw Check.mkEx("outputStreams must not be empty");
		}
		else if (outputStreams.size() == 1)
		{
			return outputStreams.get(0);
		}
		else
		{
			return new CompositeOutputStream(ImmutableList.copyOf(outputStreams));
		}
	}

	/**
	 * Convenient method to create an output stream which writes to all given files of the given directory
	 */
	public static OutputStream ofFilenames(@NonNull final Path directory, final @NonNull Set<String> filenames) throws IOException
	{
		if (filenames.isEmpty())
		{
			throw Check.mkEx("outputStreams must not be empty");
		}
		else if (filenames.size() == 1)
		{
			final File file = new File(directory.toFile(), filenames.iterator().next());
			return Files.newOutputStream(file.toPath());
		}
		else
		{
			final ImmutableList.Builder<OutputStream> outputStreams = ImmutableList.builder();
			for (final String filename : filenames)
			{
				final File file = new File(directory.toFile(), filename);
				outputStreams.add(Files.newOutputStream(file.toPath()));
			}

			return new CompositeOutputStream(outputStreams.build());
		}
	}

	@Override
	public void write(final byte @NonNull [] b) throws IOException
	{
		for (final OutputStream os : outputStreams)
		{
			os.write(b);
		}
	}

	@Override
	public void write(final byte @NonNull [] b, final int off, final int len) throws IOException
	{
		for (final OutputStream os : outputStreams)
		{
			os.write(b, off, len);
		}
	}

	@Override
	public void write(final int b) throws IOException
	{
		for (final OutputStream os : outputStreams)
		{
			os.write(b);
		}
	}

	@Override
	public void flush() throws IOException
	{
		for (final OutputStream os : outputStreams)
		{
			os.flush();
		}
	}

	@Override
	public void close() throws IOException
	{
		for (final OutputStream os : outputStreams)
		{
			os.close();
		}
	}
}
