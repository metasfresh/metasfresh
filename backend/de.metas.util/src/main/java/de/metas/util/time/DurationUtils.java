package de.metas.util.time;

import com.google.common.collect.ImmutableList;
import de.metas.util.Check;
import de.metas.util.NumberUtils;
import de.metas.util.StringUtils;
import lombok.NonNull;
import lombok.experimental.UtilityClass;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.Duration;
import java.time.temporal.ChronoUnit;
import java.time.temporal.TemporalUnit;
import java.util.List;

/*
 * #%L
 * de.metas.util
 * %%
 * Copyright (C) 2018 metas GmbH
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program. If not, see
 * <http://www.gnu.org/licenses/gpl-2.0.html>.
 * #L%
 */

@UtilityClass
public class DurationUtils
{
	private static final List<TemporalUnit> supportedTemporalUnits = ImmutableList.of(ChronoUnit.DAYS, ChronoUnit.HOURS, ChronoUnit.MINUTES, ChronoUnit.SECONDS, ChronoUnit.MILLIS);
	public static final int WORK_HOURS_PER_DAY = 8;

	/**
	 * Calculated an integer duration from a possibly non-integer source, by using a lower temporal unit (if needed). To be used in a "work" context, where a work day is equivalent to 8 hours.
	 * *<p>Example:
	 * <table border>
	 * <caption><b>Work duration conversion example</b></caption>
	 * <tr valign=top><th>Input value</th><th>Input {@link TemporalUnit}</th>
	 * <th>Output value</th><th>Output {@link TemporalUnit}</th>
	 * </tr>
	 * <tr><td align=right>0.5</td>  <td>{@link ChronoUnit#DAYS}</td><td align=right>4</td>  <td>{@link ChronoUnit#HOURS}</td></tr>
	 * <tr><td align=right>0.5</td>  <td>{@link ChronoUnit#HOURS}</td><td align=right>30</td>  <td>{@link ChronoUnit#MINUTES}</td></tr>
	 * <tr><td align=right>0.5</td>  <td>{@link ChronoUnit#MINUTES}</td><td align=right>30</td>  <td>{@link ChronoUnit#SECONDS}</td></tr>
	 * <tr><td align=right>0.5</td>  <td>{@link ChronoUnit#SECONDS}</td><td align=right>500</td>  <td>{@link ChronoUnit#MILLIS}</td></tr>
	 * </table>
	 * </p>
	 *
	 * @param duration the possibly non-integer duration value
	 * @param unit     the unit in which duration is expressed
	 * @return a {@link java.time.Duration} equivalent to the value and given time unit, but which is expressed as a Natural number.
	 * @throws RuntimeException if the given {@code duration} cannot be expressed as a natural number by using a smaller Temporal unit.
	 */
	public static Duration toWorkDuration(final @NonNull BigDecimal duration, final @NonNull TemporalUnit unit)
	{
		Check.assumeGreaterOrEqualToZero(duration, "Only positive work durations can be used");

		BigDecimal currentDuration = duration;
		TemporalUnit currentUnit = unit;

		while (NumberUtils.stripTrailingDecimalZeros(currentDuration).scale() != 0 && !currentUnit.equals(ChronoUnit.NANOS))
		{
			currentDuration = getEquivalentInSmallerTemporalUnit(currentDuration, currentUnit);
			currentUnit = supportedTemporalUnits.get(supportedTemporalUnits.indexOf(currentUnit) + 1);
		}
		final boolean longValueFound = NumberUtils.stripTrailingDecimalZeros(currentDuration).scale() == 0;
		Check.assume(longValueFound, StringUtils.formatMessage("For {} only natural numbers are supported", unit));
		final long durationLong = currentDuration.longValueExact();
		return Duration.of(durationLong, currentUnit);
	}

	public static Duration toWorkDurationRoundUp(@NonNull final BigDecimal durationBD, @NonNull final TemporalUnit unit)
	{
		return toWorkDuration(durationBD.setScale(0, RoundingMode.UP), unit);
	}

	public static BigDecimal toBigDecimal(@NonNull final Duration duration, @NonNull final TemporalUnit unit)
	{
		return BigDecimal.valueOf(toLong(duration, unit));
	}

	public static Duration fromBigDecimal(@NonNull final BigDecimal duration, @NonNull final TemporalUnit unit)
	{
		return Duration.of(duration.longValue(), unit);
	}

	public static int toInt(@NonNull final Duration duration, @NonNull final TemporalUnit unit)
	{
		return (int)toLong(duration, unit);
	}

	private static long toLong(@NonNull final Duration duration, @NonNull final TemporalUnit unit)
	{
		if (unit == ChronoUnit.SECONDS)
		{
			return duration.getSeconds();
		}
		else if (unit == ChronoUnit.MINUTES)
		{
			return duration.toMinutes();
		}
		else if (unit == ChronoUnit.HOURS)
		{
			return duration.toHours();
		}
		else if (unit == ChronoUnit.DAYS)
		{
			return duration.toDays();
		}
		else
		{
			throw Check.newException("Cannot convert " + duration + " to " + unit);
		}
	}

	private static BigDecimal getEquivalentInSmallerTemporalUnit(@NonNull final BigDecimal durationBD, @NonNull final TemporalUnit unit)
	{
		if (unit == ChronoUnit.DAYS)
		{
			return durationBD.multiply(BigDecimal.valueOf(WORK_HOURS_PER_DAY));// This refers to work hours, not calendar hours
		}
		if (unit == ChronoUnit.HOURS)
		{
			return durationBD.multiply(BigDecimal.valueOf(60));
		}
		if (unit == ChronoUnit.MINUTES)
		{
			return durationBD.multiply(BigDecimal.valueOf(60));
		}
		if (unit == ChronoUnit.SECONDS)
		{
			return durationBD.multiply(BigDecimal.valueOf(1000));
		}
		throw Check.newException("No smaller temporal unit defined for {}", unit);
	}

	public static boolean isCompleteDays(@NonNull final Duration duration)
	{
		if (duration.isZero())
		{
			return true;
		}

		final Duration daysAsDuration = Duration.ofDays(duration.toDays());
		return daysAsDuration.equals(duration);
	}
}
