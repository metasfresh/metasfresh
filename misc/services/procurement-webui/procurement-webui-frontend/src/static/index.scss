@import 'bulma/bulma.sass';
@import './variables.scss';
@import './prognose.scss';

* {
  box-sizing: border-box;
}

input:-webkit-autofill,
input:-webkit-autofill:hover,
input:-webkit-autofill:focus,
input:-webkit-autofill:active {
    transition: background-color 5000s ease-in-out 0s;
    box-shadow: 0 0 0px 1000px white inset;
}

html {
  width: 100%;
  height: 100%;
  // background-color: #f6f6f6;
  background-color: $grey-lighter;
}

body {
  color: black;
  line-height: 1.5;
  width: 100%;
  height: 100%;
}

#root {
  background-color: white;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
}

.fullsize {
  width: 100%;
  height: 100%;
}

.section {
  padding-top: 1rem;
  padding-bottom: 5.5rem;
  width: 100%;
}

.header {
  border-bottom: 1px solid $base-border;
  box-shadow: 0 0.05em 0.4em -0.125em rgba(10, 10, 10, 0.1);
  position: absolute;
  top: 0;
}

.header {
  .header-title {
    flex: 2;
    text-align: center;

    .title {
      white-space: nowrap;
      padding-right: 5px;
    }
  }

  .header-aligner {
    visibility: hidden;
  }
}

.view {
  overflow-y: auto;
  position: absolute;
  background-color: white;
  -webkit-overflow-scrolling: touch;
  transition: transform .2s ease-in-out;
}

.view-enter {
  transform: translate(-100%, 0);
}

.view-enter-active {
  transform: translate(0, 0);
}

.view-exit {
  transform: translate(-100%, 0);
}

.view-exit-active {
  transform: translate(-90%, 0);
}

.page {
  width: 100%;
  margin-top: 5.2rem;
}

.login-view {
  box-shadow: 0 0.05em 0.4em -0.125em rgba(10, 10, 10, 0.1);

  input:focus,
  input:active {
    border-color: $base-elements;
    background-color: $yellow-background;
  }
}

.login-poweredby-logo {
  position: fixed;
  bottom: 0;
  right: 1rem;
}

.daily-nav {
  width: 100%;
  color: $base-color;

    .columns {
      padding: 0rem 1rem;
      border-top: 1px solid $base-border;
      border-bottom: 1px solid $base-border;
    }

    .rows {
      display: flex;
      flex-direction: column;
      text-align: center;
      font-size: 20px;
    }

    .arrow-navigation {
      color: $base-elements;
      padding: 1rem 0;
    }
}

footer {
  min-height: 4.5rem !important;

  .link {
    display: block;
    color: $base-elements;

    svg {
      font-size: 2.25rem;
    }
  }

  .link-text {
    padding-top: 2px;
    line-height: 1;
  }
}

.back-bottom-nav {
  border-top: 1px solid $base-border;

  .nav-text {
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
  }
}

.icons-bottom-nav {
  .link {
    width: 20%;
    align-items: center;
    border-top: 1px solid $base-border;
    border-right: 1px solid $base-elements;
  }
}

.unconfirmed-count {
  position: absolute;
  color: white;
  width: 22px;
  height: 22px;
  top: 5px;
  text-align: center;
  line-height: 1rem;
  padding-top: 3px;
  right: 8px;
  padding-left: 1px;
  overflow: hidden;
  white-space: nowrap;
  border-radius: 11px;
  background-color: $danger-red;
}

.preserve-new-lines {
  white-space: pre-wrap;
}

input[type=number]::-webkit-inner-spin-button {
  cursor: pointer;
}

.product {
  margin-top: 0.4rem;
  .no-p {
    padding: 0px;
  }
  .green-check {
    color: $base-elements;
  }
}

.product-input {
  font-size: 80px;
  width: 100%;
  text-align: right;
  color: $product-input;
  font-weight: 500;
}
.product-input:focus {
  outline-color: $base-color;
  color: $base-elements;
}

.product-input::selection {
  background: $base-elements;
  color: $white-background;
}

.reset-password-form {
  .is-text {
    color: black;
  }
}

.green-color {
  color: $base-elements;
}

.is-medium.input {
  background: $yellow-background !important;
}

.week-days {
  margin-bottom: calc(9rem + 1px);
}

.bt-1 {
  border-top: 1px solid $grey-light;
}
.bb-1 {
  border-bottom: 1px solid $grey-light;
}
.br-0 {
  border-right: 0px !important;
}

.is-green-bg {
  background-color: $base-elements !important;
}

.is-green-text {
  color: $base-elements;
}

input:-webkit-autofill,
input:-webkit-autofill:hover, 
input:-webkit-autofill:focus, 
input:-webkit-autofill:active  {
    -webkit-box-shadow: 0 0 0 30px $yellow-background inset !important;
}
