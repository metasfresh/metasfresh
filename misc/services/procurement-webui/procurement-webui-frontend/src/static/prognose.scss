.prognose-box {
  position: fixed;
  bottom: 4.5rem;
  width: 100%;
  border-top: 1px solid $grey-light;
  color: $product-input;
  background: $white-background;

  svg {
    font-size: 38px;
  }
}  

/** This must be outside the prognose-box - used also in weekly report to display the trend */
.up-active {
  color: $white-background;
  background: $base-elements;
}
.up {
  color: $base-elements;
  background: $white-background;
}
.down-active {
  color: $white-background;
  background: $arrow-down;
}
.down {
  color: $arrow-down;
  background: $white-background
}
.right-active {
  color: $white-background;
  background: $arrow-right;
}
.right {
  color: $arrow-right;
  background: $white-background
}
.disabled-active {
  color: $white-background;
  background: $disabled;
}
.disabled {
  color: $disabled;
  background: $white-background
}  