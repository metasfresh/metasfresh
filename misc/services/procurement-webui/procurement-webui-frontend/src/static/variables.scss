$base-color: #000000;
$base-elements: #479F42;

$danger-red: hsl(348, 86%, 43%);
$grey-light: hsl(0, 0%, 71%);
$grey-lighter: hsl(0, 0%, 90%);
$product-input: #43423d;
$arrow-down: #9E5E2A;
$arrow-right: #1A605D;
$disabled: #9D2A2D;
$white-background: #ffffff;
$yellow-background: #FAFFBD;
$base-border: $grey-light;