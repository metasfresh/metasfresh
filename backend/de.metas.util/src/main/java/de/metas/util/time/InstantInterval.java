/*
 * #%L
 * de.metas.util
 * %%
 * Copyright (C) 2021 metas GmbH
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program. If not, see
 * <http://www.gnu.org/licenses/gpl-2.0.html>.
 * #L%
 */

package de.metas.util.time;

import lombok.Builder;
import lombok.NonNull;
import lombok.Value;

import java.time.Instant;
import java.util.Optional;

@Value
@Builder
public class InstantInterval
{
	@NonNull
	Instant from;

	@NonNull
	Instant to;

	public static InstantInterval of(@NonNull final Instant from, @NonNull final Instant to)
	{
		if (from.isAfter(to))
		{
			throw new RuntimeException("'from' cannot be after 'to'!");
		}

		return InstantInterval.builder()
				.from(from)
				.to(to)
				.build();
	}

	@NonNull
	public Optional<InstantInterval> getIntersectionWith(@NonNull final InstantInterval intervalToIntersect)
	{
		if (intervalToIntersect.getFrom().compareTo(from) >= 0 && intervalToIntersect.getTo().compareTo(to) <= 0)
		{
			return Optional.of(intervalToIntersect);
		}
		else if (intervalToIntersect.getFrom().compareTo(from) >= 0
				&& intervalToIntersect.getFrom().compareTo(to) <= 0
				&& intervalToIntersect.getTo().compareTo(to) >= 0)
		{
			return Optional.of(InstantInterval.of(intervalToIntersect.getFrom(), to));
		}
		else if (intervalToIntersect.getFrom().compareTo(from) <= 0
				&& intervalToIntersect.getTo().compareTo(to) <= 0
				&& intervalToIntersect.getTo().compareTo(from) >= 0)
		{
			return Optional.of(InstantInterval.of(from, intervalToIntersect.getTo()));
		}
		else if (intervalToIntersect.getFrom().compareTo(from) <= 0
				&& intervalToIntersect.getTo().compareTo(to) >= 0)
		{
			return Optional.of(this);
		}

		return Optional.empty();
	}
}
