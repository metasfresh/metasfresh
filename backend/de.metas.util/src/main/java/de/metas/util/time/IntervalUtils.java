/*
 * #%L
 * de.metas.util
 * %%
 * Copyright (C) 2021 metas GmbH
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program. If not, see
 * <http://www.gnu.org/licenses/gpl-2.0.html>.
 * #L%
 */

package de.metas.util.time;

import com.google.common.collect.ImmutableList;
import lombok.Builder;
import lombok.NonNull;
import lombok.Value;

import java.time.Instant;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class IntervalUtils
{
	@NonNull
	public static ImmutableList<InstantInterval> intervalDiff(
			@NonNull final ImmutableList<InstantInterval> intervals,
			@NonNull final ImmutableList<InstantInterval> intervalsToRemove)
	{
		if (intervals.isEmpty() || intervalsToRemove.isEmpty())
		{
			return intervals;
		}

		final ImmutableList<AnnotatedPoint> orderedAnnotatedPointList = buildOrderedAnnotatedPointList(intervals, intervalsToRemove);

		return removeGaps(orderedAnnotatedPointList);
	}

	@NonNull
	private static ImmutableList<AnnotatedPoint> buildOrderedAnnotatedPointList(
			@NonNull final List<InstantInterval> intervals,
			@NonNull final List<InstantInterval> intervalsToRemove)
	{
		final List<AnnotatedPoint> annotatedPoints = new ArrayList<>();

		intervals.forEach(interval -> {
			annotatedPoints.add(AnnotatedPoint.of(interval.getFrom(), AnnotatedPoint.PointType.Start));
			annotatedPoints.add(AnnotatedPoint.of(interval.getTo(), AnnotatedPoint.PointType.End));
		});

		intervalsToRemove.forEach(interval -> {
			annotatedPoints.add(AnnotatedPoint.of(interval.getFrom(), AnnotatedPoint.PointType.GapStart));
			annotatedPoints.add(AnnotatedPoint.of(interval.getTo(), AnnotatedPoint.PointType.GapEnd));
		});

		Collections.sort(annotatedPoints);

		return ImmutableList.copyOf(annotatedPoints);
	}

	@NonNull
	private static ImmutableList<InstantInterval> removeGaps(@NonNull final List<AnnotatedPoint> annotatedPointList)
	{
		final List<InstantInterval> result = new ArrayList<>();

		// iterate over the annotatedPointList
		boolean isInterval = false;
		boolean isGap = false;
		Instant intervalStart = null;

		for (final AnnotatedPoint point : annotatedPointList)
		{
			switch (point.type)
			{
				case Start:
					if (!isGap)
					{
						intervalStart = point.value;
					}
					isInterval = true;
					break;
				case End:
					if (!isGap)
					{
						//ignore the null warning as we are on the End case branch
						result.add(InstantInterval.of(intervalStart, point.value));
					}
					isInterval = false;
					break;
				case GapStart:
					if (isInterval)
					{
						result.add(InstantInterval.of(intervalStart, point.value));
					}
					isGap = true;
					break;
				case GapEnd:
					if (isInterval)
					{
						intervalStart = point.value;
					}
					isGap = false;
					break;
			}
		}

		return ImmutableList.copyOf(result);
	}

	@Value
	@Builder
	private static class AnnotatedPoint implements Comparable<AnnotatedPoint>
	{
		@NonNull
		Instant value;

		@NonNull
		PointType type;

		public static AnnotatedPoint of(@NonNull final Instant instant, @NonNull final PointType type)
		{
			return AnnotatedPoint.builder()
					.value(instant)
					.type(type)
					.build();
		}

		@Override
		public int compareTo(@NonNull final AnnotatedPoint other)
		{
			if (other.value.compareTo(this.value) == 0)
			{
				return this.type.ordinal() < other.type.ordinal() ? -1 : 1;
			}
			else
			{
				return this.value.compareTo(other.value);
			}
		}

		// the order is important here, as if multiple points have the same value
		// this is the order in which we deal with them
		public enum PointType
		{
			End, GapEnd, GapStart, Start
		}
	}
}
