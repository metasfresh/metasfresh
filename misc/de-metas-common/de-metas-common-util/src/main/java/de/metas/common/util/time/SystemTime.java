package de.metas.common.util.time;

import lombok.NonNull;

import javax.annotation.Nullable;
import java.sql.Timestamp;
import java.time.Instant;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.time.temporal.ChronoUnit;
import java.util.Calendar;
import java.util.Date;
import java.util.GregorianCalendar;

/*
 * #%L
 * de-metas-common-util
 * %%
 * Copyright (C) 2020 metas GmbH
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program. If not, see
 * <http://www.gnu.org/licenses/gpl-2.0.html>.
 * #L%
 */

public class SystemTime
{
	private static final TimeSource defaultTimeSource = new SystemTimeSource();

	@Nullable
	private static TimeSource timeSource;

	private static TimeSource getTimeSource()
	{
		return timeSource == null ? defaultTimeSource : timeSource;
	}

	/**
	 * After invocation of this method, the time returned will be the system time again.
	 */
	public static void resetTimeSource()
	{
		timeSource = null;
	}

	/**
	 * @param newTimeSource the given TimeSource will be used for the time returned by the
	 *                      methods of this class (unless it is null).
	 */
	public static void setTimeSource(@NonNull final TimeSource newTimeSource)
	{
		timeSource = newTimeSource;
	}

	public static void setFixedTimeSource(@NonNull final ZonedDateTime date)
	{
		setTimeSource(FixedTimeSource.ofZonedDateTime(date));
	}

	/**
	 * @param zonedDateTime ISO 8601 date time format (see {@link ZonedDateTime#parse(CharSequence)}).
	 *                      e.g. 2018-02-28T13:13:13+01:00[Europe/Berlin]
	 */
	public static void setFixedTimeSource(@NonNull final String zonedDateTime)
	{
		setTimeSource(FixedTimeSource.ofZonedDateTime(ZonedDateTime.parse(zonedDateTime)));
	}

	public static long millis()
	{
		return getTimeSource().millis();
	}

	public static ZoneId zoneId()
	{
		return getTimeSource().zoneId();
	}

	public static GregorianCalendar asGregorianCalendar()
	{
		final GregorianCalendar cal = new GregorianCalendar();
		cal.setTimeInMillis(millis());

		return cal;
	}

	public static Date asDate()
	{
		return new Date(millis());
	}

	public static Timestamp asTimestamp()
	{
		return new Timestamp(millis());
	}

	/**
	 * Same as {@link #asTimestamp()} but the returned date will be truncated to DAY.
	 */
	public static Timestamp asDayTimestamp()
	{
		final GregorianCalendar cal = asGregorianCalendar();
		cal.set(Calendar.HOUR_OF_DAY, 0);
		cal.set(Calendar.MINUTE, 0);
		cal.set(Calendar.SECOND, 0);
		cal.set(Calendar.MILLISECOND, 0);
		return new Timestamp(cal.getTimeInMillis());
	}

	public static Instant asInstant()
	{
		return Instant.ofEpochMilli(millis());
	}

	public static LocalDateTime asLocalDateTime()
	{
		return asZonedDateTime().toLocalDateTime();
	}

	@NonNull
	public static LocalDate asLocalDate()
	{
		return asLocalDate(zoneId());
	}

	@NonNull
	public static LocalDate asLocalDate(@NonNull final ZoneId zoneId)
	{
		return asZonedDateTime(zoneId).toLocalDate();
	}

	public static ZonedDateTime asZonedDateTime()
	{
		return asZonedDateTime(zoneId());
	}

	public static ZonedDateTime asZonedDateTimeAtStartOfDay()
	{
		return asZonedDateTime(zoneId()).truncatedTo(ChronoUnit.DAYS);
	}

	public static ZonedDateTime asZonedDateTimeAtEndOfDay(@NonNull final ZoneId zoneId)
	{
		return asZonedDateTime(zoneId)
				.toLocalDate()
				.atTime(LocalTime.MAX)
				.atZone(zoneId);
	}

	public static ZonedDateTime asZonedDateTime(@NonNull final ZoneId zoneId)
	{
		return asInstant().atZone(zoneId);
	}
}
