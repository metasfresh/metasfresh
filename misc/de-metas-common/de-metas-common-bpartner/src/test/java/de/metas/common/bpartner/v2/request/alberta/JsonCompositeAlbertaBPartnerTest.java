/*
 * #%L
 * de-metas-common-bpartner
 * %%
 * Copyright (C) 2021 metas GmbH
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program. If not, see
 * <http://www.gnu.org/licenses/gpl-2.0.html>.
 * #L%
 */

package de.metas.common.bpartner.v2.request.alberta;

import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.MapperFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.google.common.collect.ImmutableList;
import org.junit.BeforeClass;
import org.junit.Test;

import java.io.IOException;
import java.math.BigDecimal;
import java.time.Instant;
import java.time.LocalDate;

import static io.github.jsonSnapshot.SnapshotMatcher.expect;
import static io.github.jsonSnapshot.SnapshotMatcher.start;
import static org.assertj.core.api.Assertions.*;

public class JsonCompositeAlbertaBPartnerTest
{
	private final ObjectMapper mapper = new ObjectMapper()
			.findAndRegisterModules()
			.disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS)
			.disable(DeserializationFeature.ADJUST_DATES_TO_CONTEXT_TIME_ZONE)
			.enable(MapperFeature.USE_ANNOTATIONS);

	@BeforeClass
	public static void beforeAll()
	{
		start();
	}

	@Test
	public void serializeDeserialize() throws IOException
	{
		final JsonAlbertaBPartner bPartner = new JsonAlbertaBPartner();
		bPartner.setIsArchived(true);
		bPartner.setTitle("title");
		bPartner.setTitleShort("titleShort");
		bPartner.setTimestamp(Instant.parse("2019-11-22T00:00:00Z"));

		final JsonAlbertaPatient patient = new JsonAlbertaPatient();
		patient.setHospitalIdentifier("1111111");
		patient.setDischargeDate(LocalDate.parse("2019-11-22"));
		patient.setPayerIdentifier("2222222");
		patient.setPayerType("payerType");
		patient.setNumberOfInsured("333333");
		patient.setCopaymentFrom(LocalDate.parse("2021-05-24"));
		patient.setCopaymentTo(LocalDate.parse("2021-05-24"));
		patient.setIsTransferPatient(true);
		patient.setIVTherapy(true);
		patient.setFieldNurseIdentifier("4444444");
		patient.setDeactivationReason("reason");
		patient.setClassification("classification");
		patient.setCareDegree(BigDecimal.valueOf(2.2));
		patient.setDeactivationDate(LocalDate.parse("2021-05-24"));
		patient.setDeactivationComment("comment");
		patient.setCreatedAt(Instant.parse("2019-11-22T00:00:00Z"));
		patient.setCreatedByIdentifier("5555555");
		patient.setUpdatedAt(Instant.parse("2019-11-22T00:00:00Z"));
		patient.setCreatedByIdentifier("6666666");

		final JsonAlbertaCareGiver careGiver = new JsonAlbertaCareGiver();
		careGiver.setCaregiverIdentifier("11111");
		careGiver.setType("type");
		careGiver.setIsLegalCarer(true);

		final ImmutableList.Builder<JsonAlbertaCareGiver> careGivers = ImmutableList.builder();
		careGivers.add(careGiver);

		final JsonCompositeAlbertaBPartner compositeAlbertaBPartner = JsonCompositeAlbertaBPartner.builder()
				.jsonAlbertaBPartner(bPartner)
				.role(JsonBPartnerRole.Patient)
				.jsonAlbertaPatient(patient)
				.jsonAlbertaCareGivers(careGivers.build())
				.build();

		final String string = mapper.writeValueAsString(compositeAlbertaBPartner);
		assertThat(string).isNotEmpty();

		final JsonCompositeAlbertaBPartner result = mapper.readValue(string, JsonCompositeAlbertaBPartner.class);

		assertThat(result).isEqualTo(compositeAlbertaBPartner);
		expect(result).toMatchSnapshot();
	}
}
