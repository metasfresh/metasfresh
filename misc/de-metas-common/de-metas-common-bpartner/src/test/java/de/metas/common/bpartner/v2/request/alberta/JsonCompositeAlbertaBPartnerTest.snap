de.metas.common.bpartner.v2.request.alberta.JsonCompositeAlbertaBPartnerTest.serializeDeserialize=[
  {
    "jsonAlbertaBPartner" : {
      "isArchived" : true,
      "archivedSet" : true,
      "title" : "title",
      "titleSet" : true,
      "titleShort" : "titleShort",
      "titleShortSet" : true,
      "timestamp" : {
        "seconds" : 1574380800,
        "nanos" : 0
      },
      "timestampSet" : true
    },
    "role" : "Patient",
    "jsonAlbertaPatient" : {
      "hospitalIdentifier" : "1111111",
      "hospitalIdentifierSet" : true,
      "dischargeDate" : {
        "year" : 2019,
        "month" : 11,
        "day" : 22,
        "leapYear" : false
      },
      "dischargeDateSet" : true,
      "payerIdentifier" : "2222222",
      "payerIdentifierSet" : true,
      "payerType" : "payerType",
      "payerTypeSet" : true,
      "numberOfInsured" : "333333",
      "numberOfInsuredSet" : true,
      "copaymentFrom" : {
        "year" : 2021,
        "month" : 5,
        "day" : 24,
        "leapYear" : false
      },
      "copaymentFromSet" : true,
      "copaymentTo" : {
        "year" : 2021,
        "month" : 5,
        "day" : 24,
        "leapYear" : false
      },
      "copaymentToSet" : true,
      "isTransferPatient" : true,
      "transferPatientSet" : true,
      "isIVTherapy" : true,
      "ivTherapySet" : true,
      "fieldNurseIdentifier" : "4444444",
      "fieldNurseIdentifierSet" : true,
      "deactivationReason" : "reason",
      "deactivationReasonSet" : true,
      "deactivationDate" : {
        "year" : 2021,
        "month" : 5,
        "day" : 24,
        "leapYear" : false
      },
      "deactivationDateSet" : true,
      "deactivationComment" : "comment",
      "deactivationCommentSet" : true,
      "classification" : "classification",
      "classificationSet" : true,
      "careDegree" : 2.2,
      "careDegreeSet" : true,
      "createdAt" : {
        "seconds" : 1574380800,
        "nanos" : 0
      },
      "createdAtSet" : true,
      "createdByIdentifier" : "6666666",
      "createdByIdentifierSet" : true,
      "updatedAt" : {
        "seconds" : 1574380800,
        "nanos" : 0
      },
      "updatedAtSet" : true,
      "updateByIdentifierSet" : false
    },
    "jsonAlbertaCareGivers" : [
      {
        "caregiverIdentifier" : "11111",
        "type" : "type",
        "typeSet" : true,
        "isLegalCarer" : true,
        "legalCarerSet" : true
      }
    ]
  }
]